docs/topics/native/native-objc-interop.md

```diff
@@ -438,28 +438,84 @@ See more examples in the Kotlin-Swift interopedia:
 
-### NSNumber
+### Primitive types
 
 Kotlin primitive type boxes are mapped to special Swift/Objective-C classes. For example, the `kotlin.Int` box is represented
-as `KotlinInt` class instance in Swift (or `${prefix}Int` instance in Objective-C, where `prefix` is the framework names prefix).
+as the `KotlinInt` class instance in Swift (or the `${prefix}Int` instance in Objective-C, where `prefix` is the framework names prefix).
 These classes are derived from `NSNumber`, so the instances are proper `NSNumber`s supporting all corresponding operations.
 
-`NSNumber` type is not automatically translated to Kotlin primitive types when used as a Swift/Objective-C parameter type
-or return value. The reason is that `NSNumber` type doesn't provide enough information about a wrapped primitive value
+The `NSNumber` type is not automatically translated to Kotlin primitive types when used as a Swift/Objective-C parameter type
+or return value. The reason is that the `NSNumber` type doesn't provide enough information about a wrapped primitive value
 type, for example, `NSNumber` is statically not known to be `Byte`, `Boolean`, or `Double`. So Kotlin primitive values
 should be [cast to and from `NSNumber` manually](#casting-between-mapped-types).
 
-### NSMutableString
+### Strings
+
+When a Kotlin `String` is passed to Swift, it's first exported as an Objective-C object, and then the Swift compiler
+copies it one more time for a Swift conversion. This results in additional runtime overhead.
+
+To avoid that, access Kotlin strings in Swift directly as an Objective-C `NSString` instead.
+[See the conversion example](#see-the-conversion-example).
+
+#### NSMutableString
 
 `NSMutableString` Objective-C class is not available from Kotlin.
 All instances of `NSMutableString` are copied when passed to Kotlin.
 
 ### Collections
 
-Kotlin collections are converted to Swift/Objective-C collections as described in the [table above](#mappings).
-Swift/Objective-C collections are mapped to Kotlin in the same way, except for `NSMutableSet` and `NSMutableDictionary`.
+#### Kotlin -> Objective-C -> Swift
+
+When a Kotlin collection is passed to Swift, it's first converted to an Objective-C equivalent, and then the Swift compiler
+copies the entire collection and converts it into a Swift-native collection as described in the [table above](#mappings).
+
+This last conversion leads to performance costs. To prevent this, when using Kotlin collections in Swift,
+explicitly cast them to their Objective-C counterparts: `NSDictionary`, `NSArray`, or `NSSet`.
+
+##### See the conversion example {initial-collapse-state="collapsed" collapsible="true"}
+
+For example, the following Kotlin declaration:
+
+```kotlin
+data class Data(val map: Map<String, String>) {
+    fun getFromMap(key: String) = map[key]
```

Comment:
This function was only useful in my repro to compare against; if it isn’t used in this snippet, it may just be confusing to have it there.

Reply:
True, thank you! we've simplified the snippets